src/handlers/swap_quote_handlers.ts

```diff
@@ -0,0 +1,199 @@
+    private readonly _provider: SupportedProvider;
+    private readonly _swapQuoter: SwapQuoter;
+    private readonly _swapQuoteConsumer: SwapQuoteConsumer;
+    private readonly _web3Wrapper: Web3Wrapper;
+    constructor(orderbook: Orderbook, provider: SupportedProvider) {
+        this._provider = provider;
+        const swapQuoterOpts = {
+            chainId: CHAIN_ID,
+            expiryBufferMs: QUOTE_ORDER_EXPIRATION_BUFFER_MS,
+        };
+        this._swapQuoter = new SwapQuoter(this._provider, orderbook, swapQuoterOpts);
+        this._swapQuoteConsumer = new SwapQuoteConsumer(this._provider);
+        this._web3Wrapper = new Web3Wrapper(this._provider);
+    }
+    public async calculateSwapQuoteAsync(
+        buyTokenAddress: string,
+        sellTokenAddress: string,
+        buyAmount: BigNumber | undefined,
+        sellAmount: BigNumber | undefined,
+        from: string | undefined,
+        isETHSell: boolean = false,
+        slippagePercentage?: number,
+        providedGasPrice?: BigNumber,
+    ): Promise<GetSwapQuoteResponse> {
+        let swapQuote;
+        if (sellAmount !== undefined) {
+            swapQuote = await this._swapQuoter.getMarketSellSwapQuoteAsync(
+                buyTokenAddress,
+                sellTokenAddress,
+                sellAmount,
+                { slippagePercentage, gasPrice: providedGasPrice },
+            );
+        } else if (buyAmount !== undefined) {
+            swapQuote = await this._swapQuoter.getMarketBuySwapQuoteAsync(
+                buyTokenAddress,
+                sellTokenAddress,
+                buyAmount,
+                { slippagePercentage, gasPrice: providedGasPrice },
+            );
+        } else {
+            throw new Error('sellAmount or buyAmount required');
+        }
+        const {
+            makerAssetAmount,
+            totalTakerAssetAmount,
+            protocolFeeInWeiAmount: protocolFee,
+        } = swapQuote.worstCaseQuoteInfo;
+        const { orders, gasPrice } = swapQuote;
+
+        // If ETH was specified as the token to sell then we use the Forwarder
+        const extensionContractType = isETHSell ? ExtensionContractType.Forwarder : ExtensionContractType.None;
+        const {
+            calldataHexString: data,
+            ethAmount: value,
+            toAddress: to,
+        } = await this._swapQuoteConsumer.getCalldataOrThrowAsync(swapQuote, {
+            useExtensionContract: extensionContractType,
+        });
+
+        let gas;
+        if (from) {
+            gas = await this._estimateGasOrThrowRevertErrorAsync({
+                to,
+                data,
+                from,
+                value,
+                gasPrice,
+            });
+        }
+
+        const buyTokenDecimals = findTokenDecimalsIfExists(buyTokenAddress, CHAIN_ID) || DEFAULT_TOKEN_DECIMALS;
+        const sellTokenDecimals = findTokenDecimalsIfExists(sellTokenAddress, CHAIN_ID) || DEFAULT_TOKEN_DECIMALS;
```

Comment:
Should we let them optionally tell us the token decimals?

Reply:
This is only for `price`, which is a nice to have, but we definitely don't have full context over what this thing could be. We could query `decimals` on a cache miss and cache the result, this will be more important on the `/tokens/price` endpoint on how we deal with it there. The raw `assetAmounts` are returned if the caller wants to calculate the price more accurately.

I'd rather let the caller calculate it with more context rather than provide a query param for them to inject it, open to alternatives though...

Reply:
In that case should we not be returning the price at all? Would rather return no price than a bad price.